include/crm/pengine/print_ml_status.h

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2004-2018 Andrew Beekhof <andrew@beekhof.net>
+ *
+ * This source code is licensed under the GNU Lesser General Public License
+ * version 2.1 or later (LGPLv2.1+) WITHOUT ANY WARRANTY.
+ */
+
+#ifndef PENGINE_PRINT_ML_STATUS__H
+#  define PENGINE_PRINT_ML_STATUS__H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+    
+#  include <stdint.h>
+#  include <crm/pengine/status.h>
+#  include <crm/stonith-ng.h>
+
+/*
+ * Definitions indicating which items to print
+ */
+
+#define mon_show_times         (0x0001U)
+#define mon_show_stack         (0x0002U)
+#define mon_show_dc            (0x0004U)
+#define mon_show_count         (0x0008U)
+#define mon_show_nodes         (0x0010U)
+#define mon_show_resources     (0x0020U)
+#define mon_show_attributes    (0x0040U)
+#define mon_show_failcounts    (0x0080U)
+#define mon_show_operations    (0x0100U)
+#define mon_show_tickets       (0x0200U)
+#define mon_show_bans          (0x0400U)
+#define mon_show_fence_history (0x0800U)
+
+#define mon_show_headers       (mon_show_times | mon_show_stack | mon_show_dc \
+                               | mon_show_count)
+#define mon_show_default       (mon_show_headers | mon_show_nodes \
+                               | mon_show_resources)
+#define mon_show_all           (mon_show_default | mon_show_attributes \
+                               | mon_show_failcounts | mon_show_operations \
+                               | mon_show_tickets | mon_show_bans \
+                               | mon_show_fence_history)
+
+/*
+ * Definitions indicating how to output
+ */
+
+enum mon_output_format_e {
+    mon_output_none,
+    mon_output_monitor,
+    mon_output_plain,
+    mon_output_console,
+    mon_output_xml,
+    mon_output_html,
+    mon_output_cgi
+} output_format = mon_output_console;
+
+struct print_params_t
+{
+    unsigned int show; // = mon_show_default;
+    gboolean group_by_node; // = FALSE;
+    gboolean inactive_resources; // = FALSE;
+    gboolean print_timing; // = FALSE;
+    gboolean fence_history; // = FALSE
+    gboolean fence_full_history; // = FALSE
+    gboolean print_brief; // = FALSE;
+    gboolean print_pending; // = TRUE;
+    gboolean print_clone_detail; // = FALSE;
+    /* FIXME allow, detect, and correctly interpret glob pattern or regex? */
+    const char *print_neg_location_prefix; // = "";
+};
+
+/* Define exit codes for monitoring-compatible output
+ * For nagios plugins, the possibilities are
+ * OK=0, WARN=1, CRIT=2, and UNKNOWN=3
+ */
+#define MON_STATUS_WARN    CRM_EX_ERROR
+#define MON_STATUS_CRIT    CRM_EX_INVALID_PARAM
+#define MON_STATUS_UNKNOWN CRM_EX_UNIMPLEMENT_FEATURE
+
+/* Convenience macro for prettifying output (e.g. "node" vs "nodes") */
+#define s_if_plural(i) (((i) == 1)? "" : "s")
```

Comment:
I didn't know where actually to put the s_if_plural. It is used in both in print_ml_status.c and crm_mon.c, so it would be logically to put it in a header file. On the other hand s_if_plural is usually defined in .c files throughout the project. greppping it's definition will give you three other files (attrd_command.c, control_execd.c, and membership.c).

Reply:
I feel it probably belongs to one of the internal header files under include/crm/common/

Reply:
I will put it into the include/crm/common/internal.h file

Reply:
Moved to include/crm/common/internal.h

Reply:
This one was my fault. The macro was so hacky I couldn't bring myself to put it in a header. But I suppose it is helpful in more and more places, so that's probably a good idea.

For the record, there's no simple way to handle single vs plural generally. Most attempts wind up with some variant of gettext's approach, i.e. keeping a lookup table of the single and plural forms of all words or phrases that need to be handled. That seemed like overkill to me, so I went with the simple macro since all the words I needed it for are pluralized by the addition of "s".